modules/ximgproc/src/edge_drawing.cpp

```diff
@@ -12,6 +12,87 @@ namespace cv
 namespace ximgproc
 {
 
+struct ComputeGradientBody : ParallelLoopBody
+{
+void operator() (const Range& range) const CV_OVERRIDE;
+
+Mat_<uchar> src;
+mutable Mat_<ushort> gradImage;
+mutable Mat_<uchar> dirImage;
+int gradThresh;
+int op;
+bool SumFlag;
+int* grads;
+bool PFmode;
+};
+
+void ComputeGradientBody::operator() (const Range& range) const
+{
+    const int last_col = src.cols - 1;
+
+    for (int y = range.start; y < range.end; ++y)
+    {
+        const uchar* srcPrevRow = src[y - 1];
+        const uchar* srcCurRow = src[y];
+        const uchar* srcNextRow = src[y + 1];
+
+        ushort* gradRow = gradImage[y];
+        uchar* dirRow = dirImage[y];
+
+        for (int x = 1; x < last_col; ++x)
+        {
+            int com1 = srcNextRow[x + 1] - srcPrevRow[x - 1];
+            int com2 = srcPrevRow[x + 1] - srcNextRow[x - 1];
+            int gx = 0;
+            int gy = 0;
+
+            switch (op)
```

Comment:
BTW, "constant" mode switch in inner body of two loops may cause some performance loss.

(in the current patch it is just code move, so we can keep it "as is" in this PR)